### PR TITLE
GPU: Further improve shaders, textures, and blending

### PR DESCRIPTION
**Pull request by bunnei**
This is a Frankenstein PR of GPU work to [get games rendering](https://yuzu-emu.org/entry/first-games-are-running/). There are quite a few hacks in here, and some solid work. I plan on slowly breaking this up into separate more meaningful PRs. So in that light, I suggest reviewing followon PRs, not this one (this is more or less here to just provide Canary builds of games running).

### CODE REVIEW

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 1]
```diff
@@ -347,7 +357,6 @@ class GLSLGenerator {
                 LOG_CRITICAL(HW_GPU, "Unhandled instruction: 0x%02x (%s): 0x%08x",
                              static_cast<unsigned>(instr.opcode.EffectiveOpCode()),
                              OpCode::GetInfo(instr.opcode).name.c_str(), instr.hex);
-                throw DecompileFail("Unhandled instruction");
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Why did you remove this throw?

File: src/video_core/engines/shader_bytecode.h [hunk 2]
```diff
@@ -307,18 +308,28 @@ union Instruction {
...
         BitField<48, 1, u64> negate_a;
         BitField<49, 1, u64> abs_b;
         BitField<50, 1, u64> abs_d;
+        BitField<56, 1, u64> negate_imm;
+
+        float GetImm20() const {
+            u32 imm = imm20;
+            imm <<= 12;
+            imm |= negate_imm ? 0x80000000 : 0;
+            return *(f32*)&imm;
```
Comment (Subv) [inline src/video_core/engines/shader_bytecode.h]: There surely must be a better way to do this.

File: src/video_core/renderer_opengl/gl_rasterizer_cache.cpp [hunk 3]
```diff
@@ -515,6 +516,12 @@ void CachedSurface::LoadGLBuffer(VAddr load_start, VAddr load_end) {
         std::memcpy(&gl_buffer[start_offset], texture_src_data + start_offset,
                     bytes_per_pixel * width * height);
 
+        // TODO(bunnei): HACK HACK HACK - Remove before checkin!
+        u32* gl_words = reinterpret_cast<u32*>(&gl_buffer[start_offset]);
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.cpp]: The better way to do this would be to use the proper formats in GetFormatTuple

File: src/video_core/engines/shader_bytecode.h (hunk 2 above)
Comment (lioncash) [reply src/video_core/engines/shader_bytecode.h]: Yup, this is undefined behavior. It needs to be memcpy'd into a float temporary, not pointer casted.

Comment (MerryMage) [reply src/video_core/engines/shader_bytecode.h]: std::bit_cast can be used, pre-C++20 impl here: https://github.com/MerryMage/dynarmic/blob/master/src/common/cast_util.h

File: src/video_core/textures/decoders.cpp [hunk 4]
```diff
@@ -65,11 +73,21 @@ std::vector<u8> UnswizzleTexture(VAddr address, TextureFormat format, u32 width,
     std::vector<u8> unswizzled_data(width * height * bytes_per_pixel);
 
     switch (format) {
-    case TextureFormat::DXT1:
-        // In the DXT1 format, each 4x4 tile is swizzled instead of just individual pixel values.
+    case TextureFormat::BC1:
```
Comment (gdkchan) [inline src/video_core/textures/decoders.cpp]: Why not just using:
```
case TextureFormat::BC1:
case TextureFormat::BC2:
case TextureFormat::BC3:
```
with fallthrough to avoid duplicating code.

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 5]
```diff
@@ -250,55 +273,99 @@ class GLSLGenerator {
...
+                    SetDest(0, dest, "log2(" + op_a + ")", 1, 1, instr.alu.abs_d);
+                    break;
+                case SubOp::Rcp:
+                    SetDest(0, dest, "1.0 / " + op_a, 1, 1, instr.alu.abs_d);
+                    break;
+                case SubOp::Rsq:
+                    SetDest(0, dest, "inversesqrt(" + op_a + ")", 1, 1, instr.alu.abs_d);
+                    break;
+                case SubOp::Min:
+                    SetDest(0, dest, "min(" + op_a + "," + op_b + ")", 1, 1, instr.alu.abs_d);
```
Comment (gdkchan) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: nit: Space after the comma, `","` -> `", "`.

**Comment by bunnei** [comment]
Closing this - everything has either already been merged, or has a proper implementation pending review.
